src/imp/mbedtls.rs

```diff
@@ -0,0 +1,575 @@
+                } else {
+                    None
+                }
+            })
+            .ok_or(Error::Custom("No private key in pkcs12 DER".to_owned()))?
+            .map(|key| Pk::from_private_key(&key, Some(pass.as_bytes())))??;
+        let certificates: MbedtlsList<_> = pfx
+            .cert_bags(pass)?
+            .iter()
+            .map(|cert| MbedtlsCert::from_der(cert))
+            .collect::<Result<_, _>>()?;
+
+        if !certificates.is_empty() {
+            Ok(Identity {
+                key: Arc::new(key),
+                certificates: Arc::new(certificates),
+            })
+        } else {
+            Err(Error::Custom(
+                "PKCS12 file is missing certificate chain".to_owned(),
+            ))
+        }
+    }
+
+    pub fn from_pkcs8(buf: &[u8], key: &[u8]) -> Result<Identity, Error> {
+        let key = Pk::from_private_key(&key.null_terminated(), None)?;
+        let certificates = MbedtlsCert::from_pem_multiple(&buf.null_terminated())?;
+
+        if !certificates.is_empty() {
+            Ok(Identity {
+                key: Arc::new(key),
+                certificates: Arc::new(certificates),
+            })
+        } else {
+            Err(Error::Custom(
+                "X509 chain file is missing certificate chain".to_owned(),
+            ))
+        }
+    }
+
+    fn certificates(&self) -> Arc<MbedtlsList<MbedtlsCert>> {
+        self.certificates.clone()
+    }
+
+    fn private_key(&self) -> Arc<Pk> {
+        self.key.clone()
+    }
+}
+
+impl Debug for Identity {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Identity")
+            .field(
+                "certificates",
+                &self
+                    .certificates
+                    .iter()
+                    .map(|cert| cert.as_der().to_vec())
+                    .collect::<Vec<_>>(),
+            )
+            .field(
+                "key_name",
+                &self.key.name().map(String::from).map_err(Error::Tls),
+            )
+            .finish()
+    }
+}
+
+#[derive(Clone)]
+pub struct Certificate(MbedtlsBox<MbedtlsCert>);
+
+impl Certificate {
+    pub fn from_der(buf: &[u8]) -> Result<Certificate, Error> {
+        let cert = MbedtlsCert::from_der(buf).map_err(Error::Tls)?;
+        Ok(Certificate(cert))
+    }
+
+    pub fn from_pem(buf: &[u8]) -> Result<Certificate, Error> {
+        let cert = MbedtlsCert::from_pem(&buf.null_terminated()).map_err(Error::Tls)?;
+        Ok(Certificate(cert))
+    }
+
+    pub fn to_der(&self) -> Result<Vec<u8>, Error> {
+        let der = self.0.as_der().to_vec();
+        Ok(der)
+    }
+}
+
+impl Debug for Certificate {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Certificate")
+            .field(&self.0.as_der())
+            .finish()
+    }
+}
+
+pub struct TlsStream<S: Stream> {
+    ctx: Arc<Mutex<Context>>,
```

Comment:
What's the purpose of the arc and mutex?

Reply:
```rust
    is_sync::<TlsStream<TcpStream>>();
    is_send::<TlsStream<TcpStream>>();
```

Reply:
I would just flag those tests off when building against mbedtls.

Reply:
Disabled them and removed the arc and mutex. Will need to check if that is fine when using crates depending on this

Reply:
It looks like at least `ureq` depends on `TlsStream` being `Sync`. `reqwest` and `hyper-tls` doesn't build for my target anyway as they depend on `socket2`, so it's harder to check those

Reply:
All platforms should offer the same Send/Sync guarantees.

Reply:
As fortanix/rust-mbedtls/pull/262 is merged, I'll wait for release of `0.9.1` and add `Sync` support :ok_hand:

Reply:
done